From fd18a10a9f172dcc78629669ce60304924c5a2fb Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Thu, 15 Dec 2016 04:40:55 +0100
Subject: [PATCH 17/22] net: dsa: qca8k: allow swapping of mac0 and mac6

The switch allows us to swap the internal wirering of the two cpu ports.
For the HW offloading to work the ethernet MAC conencting to the LAN
ports must be wired to cpu port 0. There is HW in the wild that does not
fulfill this requirement. On these boards we need to swap the cpu ports.

Signed-off-by: John Crispin <john@phrozen.org>
---
 drivers/net/dsa/qca8k.c |   15 +++++++++++++++
 1 file changed, 15 insertions(+)

Index: linux-4.9.34/drivers/net/dsa/qca8k.c
===================================================================
--- linux-4.9.34.orig/drivers/net/dsa/qca8k.c
+++ linux-4.9.34/drivers/net/dsa/qca8k.c
@@ -518,6 +518,18 @@ qca8k_port_set_status(struct qca8k_priv
 		qca8k_reg_clear(priv, QCA8K_REG_PORT_STATUS(port), mask);
 }
 
+static void
+qca8k_exchange_mac06(struct qca8k_priv *priv, struct device_node *np)
+{
+	u32 val = qca8k_read(priv, QCA8K_REG_PORT0_PAD_CTRL);
+
+	if (of_property_read_bool(np, "qca,exchange_mac06"))
+		val |= QCA8K_PORT0_PAD_CTRL_MAC06_EXCHG;
+	else
+		val &= ~QCA8K_PORT0_PAD_CTRL_MAC06_EXCHG;
+	qca8k_write(priv, QCA8K_REG_PORT0_PAD_CTRL, val);
+}
+
 static int
 qca8k_setup(struct dsa_switch *ds)
 {
@@ -538,6 +550,9 @@ qca8k_setup(struct dsa_switch *ds)
 	if (IS_ERR(priv->regmap))
 		pr_warn("regmap initialization failed");
 
+	/* Exchange MAC0 and MAC6 */
+	qca8k_exchange_mac06(priv, priv->ds->dev->of_node);
+
 	/* Initialize CPU port pad mode (xMII type, delays...) */
 	phy_mode = of_get_phy_mode(ds->ports[ds->dst->cpu_port].dn);
 	if (phy_mode < 0) {
